accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,231 @@
+// library contracts that it requires.  If it is specified as an override, it contains the address where it has already
+// been deployed at.
+type depTreeNode struct {
+	pattern      string
+	unlinkedCode string
+	children     []*depTreeNode
+	overrideAddr *common.Address
+}
+
+// Flatten returns the subtree into a map of pattern -> unlinked contract bytecode.
+func (n *depTreeNode) Flatten() (res map[string]string) {
+	res = map[string]string{n.pattern: n.unlinkedCode}
+	for _, child := range n.children {
+		subtree := child.Flatten()
+
+		for k, v := range subtree {
+			res[k] = v
+		}
+	}
+	return res
+}
+
+// buildDepTrees is the internal version of BuildDepTrees that recursively calls itself.
+func (d *depTreeBuilder) buildDepTrees(pattern, contract string) {
+	// if the node is in the subtree set already, it has already been fully recursed/built so we can bail out.
+	if _, ok := d.subtrees[pattern]; ok {
+		return
+	}
+	node := &depTreeNode{
+		pattern:      pattern,
+		unlinkedCode: contract,
+	}
+	if addr, ok := d.overrides[pattern]; ok {
+		node.overrideAddr = &addr
+	}
+	// iterate each referenced library in the unlinked code, recurse and build its subtree.
+	reMatchSpecificPattern, err := regexp.Compile(`__\$([a-f0-9]+)\$__`)
+	if err != nil {
+		panic(err)
+	}
+	for _, match := range reMatchSpecificPattern.FindAllStringSubmatch(contract, -1) {
+		depPattern := match[1]
+		d.buildDepTrees(depPattern, d.contracts[depPattern])
+		node.children = append(node.children, d.subtrees[depPattern])
+
+		// this library can't be a root dependency if it is referenced by other contracts.
+		delete(d.roots, depPattern)
+	}
+	d.subtrees[pattern] = node
+}
+
+// BuildDepTrees will compute a set of dependency trees from a set of unlinked contracts.  The root of each tree
+// corresponds to a contract/library that is not referenced as a dependency anywhere else.  Children of each node are
+// its library dependencies.  It returns nodes that are roots of a dependency tree and nodes that aren't.
+func (d *depTreeBuilder) BuildDepTrees() (roots []*depTreeNode, nonRoots []*depTreeNode) {
+	// before the trees of dependencies are known, consider that any provided contract could be a root.
+	for pattern, _ := range d.contracts {
+		d.roots[pattern] = struct{}{}
+	}
+	for pattern, contract := range d.contracts {
+		d.buildDepTrees(pattern, contract)
+	}
+	for pattern, _ := range d.contracts {
+		if _, ok := d.roots[pattern]; ok {
+			roots = append(roots, d.subtrees[pattern])
+		} else {
+			nonRoots = append(nonRoots, d.subtrees[pattern])
+		}
+	}
+	return roots, nonRoots
+}
+
+func newDepTreeBuilder(overrides map[string]common.Address, contracts map[string]string) *depTreeBuilder {
+	return &depTreeBuilder{
+		overrides: overrides,
+		contracts: contracts,
+		subtrees:  make(map[string]*depTreeNode),
+		roots:     make(map[string]struct{}),
+	}
+}
```

Comment:
Always place the constructor right below the type definition. 

Also, I only see this being use in two places, both basically
```golang
	builder := newDepTreeBuilder(nil, contractsBins)
	roots, deps := builder.BuildDepTrees()
```
Question: is there ever any use for a builder, aside from invoking `BuildDepTrees()` on it? 
Will all usage basically be 
```golang
	root, deps := newDepTreeBuilder(overrides, contracts).BuildDepTrees()
```
?

We _could_ simplify it even more: 
```golang
func BuildDepTrees(overrides map[string]common.Address, contracts map[string]string) (roots []*depTreeNode, nonRoots []*depTreeNode) {
	return newDepTreeBuilder(overrides, contracts).buildDepTrees()
}
```

Reply:
And does `BuildDepTrees()` need to be public?

Reply:
The `depTreeBuilder` is private so I assume `BuildDepTrees` will never get exported.  But I guess in circumstances like this, the convention is to make the method name start with a lowercase (so I will change it).

Reply:
Yeah, always make everything private by default, until something forces you to make it public.